workspaces/ui-v2/src/store/MetadataProvider.tsx

```diff
@@ -0,0 +1,49 @@
+import React, { createContext, FC, useContext } from 'react';
+import { useSpectacleQuery } from '<src>/spectacle-implementations/spectacle-provider';
+import { FullPageLoader } from '<src>/optic-components/loaders';
+
+type MetadataContextValue = {
+  id: string;
+};
+
+const SpectacleMetadataQuery = `{
+  metadata {
+    id
+  }
+}`;
+
+type SpectacleMetadataResponse = {
+  metadata: {
+    id: string;
+  };
+};
+
+const MetadataContext = createContext<MetadataContextValue | null>(null);
+
+export const SpecMetadataProvider: FC = ({ children }) => {
```

Comment:
This is what we'd need to do to fix the many async calls we make + stop the `React memory leak with setState` - but I think this is what I envisage our FE state management / redux store to take the place of. The idea is we can fairly easily switch this store out.

There's also a couple of different options we have with handling loading state. In this case, we always fetch specMetadata so that the UI always has access to it and does not need to wait for loading. In other cases we could also pass down loading state and individual UI pieces could view loading state separately